src/term/term_to_net.rs

```diff
@@ -129,125 +160,126 @@ impl EncodeTermState<'_> {
     })
   }
 
-  fn link(&mut self, x: Trg, y: Trg) {
-    match (x, y) {
-      (Trg::Port(p), Trg::Port(q)) => self.inet.link(p, q),
-      (Trg::Var(v), t) | (t, Trg::Var(v)) => match &mut self.vars[v] {
-        x @ None => *x = Some(t),
-        x => {
-          let u = x.take().unwrap();
-          self.link(t, u);
+  fn link(&mut self, a: Place<'t>, b: Place<'t>) {
+    match (a, b) {
+      (Place::Tree(a), Place::Tree(b)) => self.redexes.push(LoanedMut::merge(Default::default(), |r, m| {
+        m.place(b, &mut r.0);
+        m.place(a, &mut r.1);
+      })),
+      (Place::Tree(t), Place::Hole(h)) | (Place::Hole(h), Place::Tree(t)) => {
+        t.place(h);
+      }
+      (Place::Hole(a), Place::Hole(b)) => {
+        let var = Tree::Var { nam: num_to_name(self.name_idx) };
+        self.name_idx += 1;
+        *a = var.clone();
+        *b = var;
+      }
+      (Place::Wire(v), p) | (p, Place::Wire(v)) => {
+        let v = &mut self.wires[v];
+        match v.take() {
+          Some(q) => self.link(p, q),
+          None => *v = Some(p),
         }
-      },
+      }
     }
   }
 
+  fn new_ctr(&mut self, kind: CtrKind) -> (Place<'t>, Place<'t>, Place<'t>) {
+    self.created_nodes += 1;
+    let (ports, node) =
+      LoanedMut::loan_with(Tree::Ctr { lab: kind.to_lab(), ports: vec![Tree::Era, Tree::Era] }, |t, l| {
+        let Tree::Ctr { ports, .. } = t else { unreachable!() };
+        l.loan_mut(ports)
+      });
+    let (a, b) = ports.split_at_mut(1);
+    (Place::Tree(node), Place::Hole(&mut a[0]), Place::Hole(&mut b[0]))
+  }
+
   /// Adds a list-like tree of nodes of the same kind to the inet.
   fn make_node_list(
     &mut self,
-    kind: NodeKind,
-    mut up: Trg,
-    mut els: impl DoubleEndedIterator<Item = impl FnOnce(&mut Self, Trg)>,
+    kind: CtrKind,
+    mut up: Place<'t>,
+    mut els: impl DoubleEndedIterator<Item = impl FnOnce(&mut Self, Place<'t>)>,
   ) {
     let last = els.next_back().unwrap();
     for item in els {
-      let node = self.inet.new_node(kind.clone());
-      self.link(up, Trg::Port(Port(node, 0)));
-      item(self, Trg::Port(Port(node, 1)));
-      up = Trg::Port(Port(node, 2));
+      let node = self.new_ctr(kind);
+      self.link(up, node.0);
+      item(self, node.1);
+      up = node.2;
     }
     last(self, up);
   }
 
-  fn new_var(&mut self) -> usize {
-    let i = self.vars.len();
-    self.vars.push(None);
+  fn new_wire(&mut self) -> usize {
+    let i = self.wires.len();
+    self.wires.push(None);
     i
   }
 
-  fn fan_kind(&mut self, fan: &FanKind, tag: &Tag) -> NodeKind {
+  fn fan_kind(&mut self, fan: &FanKind, tag: &Tag) -> CtrKind {
     let lab = self.labels[*fan].generate(tag);
-    Ctr(if *fan == FanKind::Tup { Tup(lab) } else { Dup(lab.unwrap()) })
+    if *fan == FanKind::Tup { Tup(lab) } else { Dup(lab.unwrap()) }
   }
 
-  fn link_global(&mut self, name: &Name, trg: Trg) {
+  fn link_global(&mut self, name: &Name, place: Place<'t>) {
```

Comment:
Do you think making scoped binds share the same logic as scopeless ones would be an improvement?
`link_global` is just a `half_link` and the only thing that differentiates scoped from scopeless at this point is the different namespaces.
With this PR you use the fact that scoped vars are scoped to always only link when the var use is encountered, but this extra bit of information is not really necessary.

Reply:
Hmm, could it be changed to share the same logic? I was under the impression that this algorithm had to account for shadowing (hence the `Vec`), so an approach like used for global vars would break on things like `@a ((@a a), a)`.

Reply:
If it doesn't actually need to account for shadowing then yeah I think it'd be nicer to use the same logic.

Reply:
At this point we've renamed all the variables to have unique names